src/OpenLoco/StationManager.cpp

```diff
@@ -124,22 +128,210 @@ namespace OpenLoco::StationManager
     // 0x048F988
     string_id generateNewStationName(StationId_t stationId, TownId_t townId, Map::Pos3 position, uint8_t mode)
     {
-        auto* station = get(stationId);
-        if (station == nullptr)
-            return StringIds::null;
-
-        station->name = StringIds::null;
-
-        registers regs;
-        regs.esi = reinterpret_cast<int32_t>(station);
-        regs.ebx = townId;
-        regs.dh = static_cast<uint8_t>(position.z / 4);
-        regs.dl = mode;
-        regs.ax = position.x & 0xFFE0;
-        regs.cx = position.y & 0xFFE0;
-
-        call(0x048F988, regs);
-        return regs.bx;
+        // Bit mask for station names already used in the current town.
+        uint32_t realNamesInUse{};    // ebp
+        uint32_t ordinalNamesInUse{}; // edi
+        for (auto& station : stations())
+        {
+            if (station.empty())
+                continue;
+
+            if (!StringManager::isTownName(station.name))
+                continue;
+
+            auto nameKey = StringManager::fromTownName(station.name) - StringIds::station_town;
+            if (nameKey >= 47)
+                continue;
+
+            if (station.town != townId)
+                continue;
+
+            if (nameKey < 27)
+                realNamesInUse |= (1 << nameKey);
+            else
+                ordinalNamesInUse |= (1 << (nameKey - 27));
+        }
+
+        if (mode == 1)
+        {
+            // Airport
+            if ((realNamesInUse & (1 << 13)) == 0)
```

Comment:
can we add constants for all these flags. With a note that they must follow the same order as `StringIds::station_town_north...`